### ISSUE TITLE
Seemingly valid truss generates "LinAlgError: Singular matrix"

### ISSUE THREAD

**Issue by dknesek** [cmccomb/truss-me#6]
My son has written a wrapper that iterates through different joint coordinates (of J[3], J[4], and J[5]) as an optimization problem. Although many execute fine, there are seemingly valid trusses that blow up with a singular matrix error. Can you help?

J   0.0 0.0 0.0 0.0 0.0 1.0
J   0.18    0.025   0.0 0.0 0.0 1.0
J   0.36    0.05    0.0 0.0 0.0 1.0
J   0.05    0.05    0.0 0.0 0.0 1.0
J   0.14    0.1 0.0 0.0 0.0 1.0
J   0.26    0.14    0.0 0.0 0.0 1.0
M   0   1   BALSA   bar t=0.0   r=0.002 
M   1   2   BALSA   bar t=0.0   r=0.002 
M   0   3   BALSA   bar t=0.0   r=0.002 
M   3   4   BALSA   bar t=0.0   r=0.002 
M   4   5   BALSA   bar t=0.0   r=0.002 
M   5   2   BALSA   bar t=0.0   r=0.002 
M   3   1   BALSA   bar t=0.0   r=0.002 
M   4   1   BALSA   bar t=0.0   r=0.002 
M   5   1   BALSA   bar t=0.0   r=0.002 
L   1   0.0 -2.0    0.0

"BALSA":   {"rho": 150, "E":   3.71_pow(10, 9), "Fy":  1_pow(10, 6)},

**Comment by cmccomb** [comment]
Unfortunately, I wasn't able to recreate your problem. I have a guess though - the configuration that you pasted doesn't have any supports in the right direction. Supports are defined in the last three numbers on every J line. Those numbers essentially define the degrees of freedom of the joint. A line ending in "0.0 0.0 1.0" indicates that the joint can deflect in the X and Y direction, but not Z. A line ending in "1.0 1.0 1.0" means that the joints isn't allowed to deflect in any direction. Because you're pushing against the truss in the Y-direction, you need to allow the truss to push back in the same direction. In other words, you need to change two of you joints to be "0.0 1.0 1.0". Changing one of those to "1.0 1.0 1.0" probably wouldn't hurt either (mostly for numerical stability).

TL;DR:
Try changing your first three lines to this:
J 0.0 0.0 0.0 1.0 1.0 1.0
J 0.18 0.025 0.0 0.0 0.0 1.0
J 0.36 0.05 0.0 0.0 1.0 1.0

Please let me know if this helps! Sounds like a cool project.

**Comment by dknesek** [comment]
Thank you for the fast reply.

Does deflect mean "move"?

He's trying to simulate a balsa wood bridge resting across an open gap, held up by a test stand on both ends. Neither end is pinned. But because both ends are suspended by the stand, they are supported in the Y dimension, correct? So the Y value should be 1.0 to indicate that the joints can't deflect in the Y direction, correct? He shouldn't have to do anything with Z if everything is 2D, right?

Also, yield means stretch to the point of rupturing, right. I think I know what buckling means. What we don't understand is what is meant by Axial-force(kN),  FOS-yielding, FOS-buckling. We think that Axial force is the force of compression or tension in the member, but aren't sure what use the other two are.

Thank you again.

**Comment by cmccomb** [comment]
Yes, deflect means move. Sorry for not being clear!

You're correct about the ends being supported in the y-dimension, so they should have 1.0 for Y. You should still leave all of the Z values as 1.0, because this makes sure that all the deflection happens in the XY plane (numerical round off can lead to deflection in the Z direction, even if there aren't forces in that direction). Setting the X value to 0 for one of the ends won't change the results, but should help with the stability of the solution.

When you're dealing with metals like steel, yielding is the point at which it begins to stretch irreversibly, and rupture is when the material breaks in two (a separate event that happens well after yielding). When you're dealing with wood, yielding and rupture happen at essentially the same time. This is why you can bend a paperclip back and forth, but a toothpick just snaps. So, since you're dealing with balsa, yielding will give you the point at which the bridge falls apart.

You're right about axial-force(kN) - its the force of compression or tension in the member, measured in kilo-Newtons. FOS-yielding is the Factor Of Safety against yielding - the total force that the member can sustain before it yields, divided by the force being applied to it. FOS-buckling is the same idea - the total force that the member can sustain before it buckles, divided by the force being applied to it. Both of these should be greater than 1.0 for a successful design.

Please let me know if you have any other questions!

**Comment by dknesek** [comment]
That seems to have done the trick. It didn't work until I added the `d=2` in the `joint.roller(d=2)` invocation.

Thank you much. One more question - why do you multiply the forces by 100 billion on line 50 of evaluate.py?

`forces *= pow(10, 10)`

**Comment by cmccomb** [comment]
That only occurs if the condition number is very large (see the if statement on the previous line). If that statement is true, then the truss has some sort of issue (can't be solved, not stable, etc.). Multiplying the forces by pow(10, 10) is a holdover from a project that I previously used the code for - a better option would probably be to throw an error if the condition number is large. I'll try to update that later today.

**Comment by dknesek** [comment]
One more question, I hope.

I'm trying to enter the correct properties for balsa. Is "Tensile Strength" of wood essentially the same as your "Yeild Stength"? The reason I ask is that a lot of the literature that documents the mechanical properties of wood don't capture Yeild Strenth, but capture Tensile Strength. If not, is there a good way of converting? (Like saying that Yeild Strength is ~80% of Tensile Strength?)

**Comment by cmccomb** [comment]
Yes, for these purposes yield strength is synonymous with tensile strength. But for some kinds of wood the tensile strength can be different than the compression strength (http://www.fpl.fs.fed.us/documnts/fplgtr/fplgtr113/ch04.pdf). 

Also, a hint for balsa wood bridges: most end up failing at the joints.

**Comment by dknesek** [comment]
Does your software assume that yield strength and compression strength are identical?

**Comment by cmccomb** [comment]
Yes it does - this is usually true for metals, which was my original intent when creating the software. 

If you're interested in modifying it, you could do so in the set_force method of the Member class. I don't know how time-sensitive the project is, but I probably won't be able to make that fix for a week or so.

**Comment by dknesek** [comment]
The competition is next weekend. You have already been a tremendous help.

If you can point me to the equations to read about to make the change, I can probably make these myself. I will probably hard code the values for balsa, rather than make it a config parameter. But after I'm done, I would be willing to do it right and submit a patch back to you.

It seems like you must have the equations already in there. You're just using the same variable in both equations, and all I would have to do is substitute a new one in the compression strength locations.

**Comment by cmccomb** [comment]
The equations are pretty straight-forward. Line 188 in member.py is:

```
self.fos_yielding = self.Fy/abs(self.force/self.area)
```

This basic equation applies for both tension and compression. Its just a matter of checking the sign of self.force (positive for tension, negative for compression), and applying a different value for self.Fy depending on compression/tension.

**Comment by dknesek** [comment]
excellent

**Comment by cmccomb** [comment]
Closing this issue - please let me know if you need anything else.
